streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java

```diff
@@ -0,0 +1,198 @@
+import org.apache.kafka.streams.state.internals.ThreadCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * This is the thread responsible for keeping all Global State Stores updated.
+ * It delegates most of the responsibility to the internal class StateConsumer
+ */
+public class GlobalStreamThread extends Thread {
+
+    private static final Logger log = LoggerFactory.getLogger(GlobalStreamThread.class);
+    private final StreamsConfig config;
+    private final Consumer<byte[], byte[]> consumer;
+    private final StateDirectory stateDirectory;
+    private final Time time;
+    private final ThreadCache cache;
+    private final StreamsMetrics streamsMetrics;
+    private final ProcessorTopology topology;
+    private volatile boolean running = false;
+    private volatile StreamsException startupException;
+
+    public GlobalStreamThread(final ProcessorTopology topology,
+                              final StreamsConfig config,
+                              final Consumer<byte[], byte[]> globalConsumer,
+                              final StateDirectory stateDirectory,
+                              final Metrics metrics,
+                              final Time time,
+                              final String clientId
+    ) {
+        super("GlobalStreamThread");
+        this.topology = topology;
+        this.config = config;
+        this.consumer = globalConsumer;
+        this.stateDirectory = stateDirectory;
+        this.time = time;
+        long cacheSizeBytes = Math.max(0, config.getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG) /
+                (config.getInt(StreamsConfig.NUM_STREAM_THREADS_CONFIG) + 1));
+        final String threadClientId = clientId + "-" + getName();
+        this.cache = new ThreadCache(threadClientId, cacheSizeBytes, new ThreadCache.NullThreadCacheMetrics());
+        this.streamsMetrics = new StreamThread.StreamsMetricsImpl(metrics, threadClientId);
+    }
+
+    static class StateConsumer {
+        private final Consumer<byte[], byte[]> consumer;
+        private final GlobalStateMaintainer stateMaintainer;
+        private final Time time;
+        private final long pollMs;
+        private final long flushInterval;
+
+        private long lastFlush;
+
+        StateConsumer(final Consumer<byte[], byte[]> consumer,
+                      final GlobalStateMaintainer stateMaintainer,
+                      final Time time,
+                      final long pollMs,
+                      final long flushInterval) {
+            this.consumer = consumer;
+            this.stateMaintainer = stateMaintainer;
+            this.time = time;
+            this.pollMs = pollMs;
+            this.flushInterval = flushInterval;
+        }
+
+        void initialize() {
+            final Map<TopicPartition, Long> partitionOffsets = stateMaintainer.initialize();
+            consumer.assign(partitionOffsets.keySet());
```

Comment:
As the restore consumer is shared, can we make sure no parallel restore operation messes with global state thread here? Applies to the whole class...

Also -- if we change partitions assignment -- would we need to restore a previous assignment or maybe better just extend the current assignment (instead or replacing it)? Or does every "user" of the consumer restores its own assignment each time it uses the consumer (this might also imply that doing the assignment in `initialize` is wrong)?


Btw: how does state recovery work today? single threaded?